Disabling CAPSTONE_RISCV_SUPPORT causes a crash during the build

<!-- This template is meant for GENERAL bug reports.
For bugs regarding incorrect disassembly,
please use the "Bug report - Incorrect disassembly" template.

Please be as descriptive as possible
-->

### Work environment

<!-- Filling this table is mandatory -->

| Questions                                | Answers
|------------------------------------------|--------------------
| System Capstone runs on OS/arch/bits     | Debian x86_64
| Capstone module affected                 | All, but RISC-V
| Source of Capstone                       | `git clone`
| Version/git commit                       | v6.0.0, 08e655c2b76b02028569d29df01e6f0022734924

<!-- OTHER BUGS -->

### Expected behavior
Disabling `CAPSTONE_RISCV_SUPPORT` from being built shouldn't cause an error.

### Actual behavior

Excluding it results in an error due to `getSysRegName()` in cstool/cstool_riscv.c and suite/cstest/src/test_detail_riscv.c.

### Steps to reproduce the behavior

```bash
cmake -B build -DCMAKE_BUILD_TYPE=Release -DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_X86_SUPPORT=1
cmake --build build -j
```

<!-- ADDITIONAL CONTEXT -->

### Additional Logs, screenshots, source code,  configuration dump, ...

CMake error:
```bash
/usr/bin/riscv64-linux-gnu-ld.bfd: CMakeFiles/cstool.dir/cstool/cstool_riscv.c.o: in function `print_insn_detail_riscv':
/root/capstone/cstool/cstool_riscv.c:47:(.text+0x170): undefined reference to `getSysRegName'
/usr/bin/riscv64-linux-gnu-ld.bfd: relocation R_RISCV_CALL_PLT against `getSysRegName' which may bind externally can not be used when making a shared object; recompile with -fPIC
```

--- a/.github/workflows/cstool.yml
+++ b/.github/workflows/cstool.yml
@@ -284,5 +284,6 @@ jobs:
         insn_bytes: ${{ matrix.config.insn_bytes }}
         cs_arch: ${{ matrix.config.cs_arch }}
       run: |
+        cstool -h
         cstool -d ${cs_arch} ${insn_bytes}
 

--- a/.github/workflows/cstool.yml
+++ b/.github/workflows/cstool.yml
@@ -1,0 +1,288 @@
+name: cstool
+on:
+  push:
+    paths-ignore:
+      - ".gitignore"
+      - "docs/**"
+      - "ChangeLog"
+      - "CREDITS.TXT"
+      - "COMPILE_MAKE.TXT"
+      - "BUILDING.md"
+      - "CONTRIBUTING.md"
+      - "LICENSE.TXT"
+      - "LICENSE_LLVM.TXT"
+      - "README.md"
+      - "RELEASE_NOTES"
+      - "SPONSORS.TXT"
+      - "TODO"
+  pull_request:
+
+# Stop previous runs on the same branch on new push
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+env:
+  CI: true
+  UBSAN_OPTIONS: "halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1"
+  ASAN_OPTIONS: "halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1"
+  LSAN_OPTIONS: "halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1"
+
+jobs:
+  Linux:
+    runs-on: ${{ matrix.config.os }}
+    name: ${{ matrix.config.name }}
+    strategy:
+      fail-fast: false
+      matrix:
+        config:
+          - {
+              name: 'CAPSTONE_X86_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_X86_SUPPORT=1',
+              insn_bytes: '8d4c320801d8',
+              cs_arch: 'x32'
+            }
+          - {
+              name: 'CAPSTONE_ARM_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_ARM_SUPPORT=1',
+              insn_bytes: 'f0ff2005',
+              cs_arch: 'arm+thumb'
+            }
+          - {
+              name: 'CAPSTONE_AARCH64_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_AARCH64_SUPPORT=1',
+              insn_bytes: '2090226f',
+              cs_arch: 'aarch64'
+            }
+          - {
+              name: 'CAPSTONE_MIPS_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_MIPS_SUPPORT=1',
+              insn_bytes: '0c10009700000000',
+              cs_arch: 'mips64r6'
+            }
+          - {
+              name: 'CAPSTONE_PPC_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_PPC_SUPPORT=1',
+              insn_bytes: '43200c074156ff17',
+              cs_arch: 'ppc64be'
+            }
+          - {
+              name: 'CAPSTONE_SPARC_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_SPARC_SUPPORT=1',
+              insn_bytes: 'bfb00c20',
+              cs_arch: 'sparc'
+            }
+          - {
+              name: 'CAPSTONE_SYSTEMZ_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_SYSTEMZ_SUPPORT=1',
+              insn_bytes: '1a00000000ed',
+              cs_arch: 'systemz_arch14'
+            }
+          - {
+              name: 'CAPSTONE_XCORE_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_XCORE_SUPPORT=1',
+              insn_bytes: 'fe0ffe171317',
+              cs_arch: 'xcore'
+            }
+          - {
+              name: 'CAPSTONE_M68K_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_M68K_SUPPORT=1',
+              insn_bytes: 'f010f000',
+              cs_arch: 'm68k'
+            }
+          - {
+              name: 'CAPSTONE_TMS320C64X_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_TMS320C64X_SUPPORT=1',
+              insn_bytes: '01ac884081ac884300000000029032960280469e053c83e60b0c8b24',
+              cs_arch: 'tms320c64x'
+            }
+          - {
+              name: 'CAPSTONE_M680X_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_M680X_SUPPORT=1',
+              insn_bytes: '6b100071100072101039',
+              cs_arch: 'm6800'
+            }
+          - {
+              name: 'CAPSTONE_EVM_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_EVM_SUPPORT=1',
+              insn_bytes: '606150',
+              cs_arch: 'evm'
+            }
+          - {
+              name: 'CAPSTONE_WASM_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_WASM_SUPPORT=1',
+              insn_bytes: '2000',
+              cs_arch: 'wasm'
+            }
+          - {
+              name: 'CAPSTONE_MOS65XX_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_MOS65XX_SUPPORT=1',
+              insn_bytes: 'a112a512',
+              cs_arch: '6502'
+            }
+          - {
+              name: 'CAPSTONE_BPF_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_BPF_SUPPORT=1',
+              insn_bytes: 'c6c70fcc431fb9f5',
+              cs_arch: 'ebpf'
+            }
+          - {
+              name: 'CAPSTONE_RISCV_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_RISCV_SUPPORT=1',
+              insn_bytes: '57144a14',
+              cs_arch: 'riscv64+v+fd'
+            }
+          - {
+              name: 'CAPSTONE_SH_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_SH_SUPPORT=1',
+              insn_bytes: '3211920032493100 ',
+              cs_arch: 'sh2a-fpu'
+            }
+          - {
+              name: 'CAPSTONE_TRICORE_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_TRICORE_SUPPORT=1',
+              insn_bytes: '17016002',
+              cs_arch: 'tc180'
+            }
+          - {
+              name: 'CAPSTONE_ALPHA_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_ALPHA_SUPPORT=1',
+              insn_bytes: 'c3002248',
+              cs_arch: 'alpha'
+            }
+          - {
+              name: 'CAPSTONE_HPPA_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_HPPA_SUPPORT=1',
+              insn_bytes: 'a04137e5',
+              cs_arch: 'hppa20be'
+            }
+          - {
+              name: 'CAPSTONE_LOONGARCH_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_LOONGARCH_SUPPORT=1',
+              insn_bytes: '40b2aa74',
+              cs_arch: 'loongarch64'
+            }
+          - {
+              name: 'CAPSTONE_XTENSA_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_XTENSA_SUPPORT=1',
+              insn_bytes: 'f60900',
+              cs_arch: 'esp8266'
+            }
+          - {
+              name: 'CAPSTONE_ARC_SUPPORT only',
+              os: ubuntu-latest,
+              diet-build: 'OFF',
+              enable-asan: 'ON',
+              build_type: 'Debug',
+              build_options: '-DCAPSTONE_ARCHITECTURE_DEFAULT=0 -DCAPSTONE_ARC_SUPPORT=1',
+              insn_bytes: '3178',
+              cs_arch: 'arc'
+            }
+
+    steps:
+    - uses: actions/checkout@v4
+
+    - name: cmake
+      env:
+        asan: ${{ matrix.config.enable-asan }}
+        build_options: ${{ matrix.config.build_options }}
+        build_type: ${{ matrix.config.build_type }}
+      run: |
+        cmake -DCMAKE_BUILD_TYPE=${build_type} -DCAPSTONE_INSTALL=1 -DCAPSTONE_BUILD_SHARED_LIBS=1 -DCMAKE_INSTALL_PREFIX=/usr -DENABLE_ASAN=${asan} ${build_options} -B build .
+        sudo cmake --build build --config ${build_type} --target install
+
+    - name: Run cstest 
+      env:
+        insn_bytes: ${{ matrix.config.insn_bytes }}
+        cs_arch: ${{ matrix.config.cs_arch }}
+      run: |
+        cstool -d ${cs_arch} ${insn_bytes}
+

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -242,6 +242,9 @@ set(HEADERS_COMMON
 ## architecture support
 if(CAPSTONE_ARM_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_ARM)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_arm.c)
     set(SOURCES_ARM
         arch/ARM/ARMBaseInfo.c
         arch/ARM/ARMDisassembler.c
@@ -272,6 +275,9 @@ endif()
 
 if(CAPSTONE_AARCH64_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_AARCH64)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_aarch64.c)
     set(SOURCES_AARCH64
         arch/AArch64/AArch64BaseInfo.c
         arch/AArch64/AArch64Disassembler.c
@@ -301,6 +307,9 @@ endif()
 
 if(CAPSTONE_MIPS_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_MIPS)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_mips.c)
     set(SOURCES_MIPS
         arch/Mips/MipsDisassembler.c
         arch/Mips/MipsInstPrinter.c
@@ -332,6 +341,9 @@ endif()
 
 if(CAPSTONE_PPC_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_POWERPC)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_powerpc.c)
     set(SOURCES_PPC
         arch/PowerPC/PPCDisassembler.c
         arch/PowerPC/PPCInstPrinter.c
@@ -360,6 +372,9 @@ endif()
 
 if(CAPSTONE_X86_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_X86)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_x86.c)
     set(SOURCES_X86
         arch/X86/X86Disassembler.c
         arch/X86/X86DisassemblerDecoder.c
@@ -413,6 +428,9 @@ endif()
 
 if(CAPSTONE_SPARC_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_SPARC)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_sparc.c)
     set(SOURCES_SPARC
         arch/Sparc/SparcDisassembler.c
         arch/Sparc/SparcDisassemblerExtension.c
@@ -444,6 +462,9 @@ endif()
 
 if(CAPSTONE_SYSTEMZ_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_SYSTEMZ)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_systemz.c)
     set(SOURCES_SYSTEMZ
         arch/SystemZ/SystemZDisassembler.c
         arch/SystemZ/SystemZDisassemblerExtension.c
@@ -475,6 +496,9 @@ endif()
 
 if(CAPSTONE_XCORE_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_XCORE)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_xcore.c)
     set(SOURCES_XCORE
         arch/XCore/XCoreDisassembler.c
         arch/XCore/XCoreInstPrinter.c
@@ -495,6 +519,9 @@ endif()
 
 if(CAPSTONE_M68K_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_M68K)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_m68k.c)
     set(SOURCES_M68K
         arch/M68K/M68KDisassembler.c
         arch/M68K/M68KInstPrinter.c
@@ -507,6 +534,9 @@ endif()
 
 if(CAPSTONE_TMS320C64X_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_TMS320C64X)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_tms320c64x.c)
     set(SOURCES_TMS320C64X
         arch/TMS320C64x/TMS320C64xDisassembler.c
         arch/TMS320C64x/TMS320C64xInstPrinter.c
@@ -526,6 +556,9 @@ endif()
 
 if(CAPSTONE_M680X_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_M680X)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_m680x.c)
     set(SOURCES_M680X
         arch/M680X/M680XDisassembler.c
         arch/M680X/M680XInstPrinter.c
@@ -540,6 +573,9 @@ endif()
 
 if(CAPSTONE_EVM_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_EVM)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_evm.c)
     set(SOURCES_EVM
         arch/EVM/EVMDisassembler.c
         arch/EVM/EVMInstPrinter.c
@@ -556,6 +592,9 @@ endif()
 
 if(CAPSTONE_WASM_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_WASM)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_wasm.c)
     set(SOURCES_WASM
         arch/WASM/WASMDisassembler.c
         arch/WASM/WASMInstPrinter.c
@@ -571,6 +610,9 @@ endif()
 
 if(CAPSTONE_MOS65XX_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_MOS65XX)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_mos65xx.c)
     set(SOURCES_MOS65XX
             arch/MOS65XX/MOS65XXModule.c
             arch/MOS65XX/MOS65XXDisassembler.c)
@@ -581,6 +623,9 @@ endif()
 
 if(CAPSTONE_BPF_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_BPF)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_bpf.c)
     set(SOURCES_BPF
         arch/BPF/BPFDisassembler.c
         arch/BPF/BPFInstPrinter.c
@@ -598,6 +643,9 @@ endif()
 
 if(CAPSTONE_RISCV_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_RISCV)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_riscv.c)
     set(SOURCES_RISCV
         arch/RISCV/RISCVBaseInfo.c
         arch/RISCV/RISCVDisassembler.c
@@ -636,6 +684,9 @@ endif()
 
 if(CAPSTONE_SH_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_SH)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_sh.c)
     set(SOURCES_SH
         arch/SH/SHDisassembler.c
         arch/SH/SHInstPrinter.c
@@ -651,6 +702,9 @@ endif()
 
 if (CAPSTONE_TRICORE_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_TRICORE)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_tricore.c)
     set(SOURCES_TRICORE
             arch/TriCore/TriCoreDisassembler.c
             arch/TriCore/TriCoreInstPrinter.c
@@ -671,6 +725,9 @@ endif ()
 
 if (CAPSTONE_ALPHA_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_ALPHA)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_alpha.c)
     set(SOURCES_ALPHA
             arch/Alpha/AlphaDisassembler.c
             arch/Alpha/AlphaInstPrinter.c
@@ -694,6 +751,9 @@ endif ()
 
 if(CAPSTONE_HPPA_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_HPPA)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_hppa.c)
     set(SOURCES_HPPA
         arch/HPPA/HPPADisassembler.c
         arch/HPPA/HPPAInstPrinter.c
@@ -711,6 +771,9 @@ endif()
 
 if (CAPSTONE_LOONGARCH_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_LOONGARCH)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_loongarch.c)
     set(SOURCES_LOONGARCH
             arch/LoongArch/LoongArchDisassembler.c
             arch/LoongArch/LoongArchDisassemblerExtension.c
@@ -728,6 +791,9 @@ endif ()
 
 if(CAPSTONE_XTENSA_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_XTENSA)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_xtensa.c)
     set(SOURCES_XTENSA
         arch/Xtensa/XtensaDisassembler.c
         arch/Xtensa/XtensaInstPrinter.c
@@ -744,6 +810,9 @@ endif()
 
 if (CAPSTONE_ARC_SUPPORT)
     add_definitions(-DCAPSTONE_HAS_ARC)
+    set(CSTOOL_SRC
+        ${CSTOOL_SRC}
+        cstool/cstool_arc.c)
     set(SOURCES_ARC
             arch/ARC/ARCDisassembler.c
             arch/ARC/ARCInstPrinter.c
@@ -1005,7 +1074,7 @@ if(CAPSTONE_INSTALL)
 endif()
 
 if(CAPSTONE_BUILD_CSTOOL)
-    file(GLOB CSTOOL_SRC cstool/*.c)
+    set(CSTOOL_SRC ${CSTOOL_SRC} cstool/getopt.c cstool/cstool.c)
     add_executable(cstool ${CSTOOL_SRC})
     target_link_libraries(cstool PRIVATE capstone)
 

--- a/cstool/cstool.c
+++ b/cstool/cstool.c
@@ -620,6 +620,16 @@ static const char *get_arch_name(cs_arch arch)
 	}
 }
 
+static inline bool contains_supported(cs_arch needles[CS_ARCH_MAX])
+{
+	for (size_t i = 0; i < CS_ARCH_MAX && needles[i] != CS_ARCH_MAX; i++) {
+		if (cs_support(needles[i])) {
+			return true;
+		}
+	}
+	return false;
+}
+
 static void usage(char *prog)
 {
 	int i, j;
@@ -638,6 +648,9 @@ static void usage(char *prog)
 
 	printf("\nArch specific options:\n");
 	for (i = 0; all_opts[i].name; i++) {
+		if (!contains_supported(all_opts[i].archs)) {
+			continue;
+		}
 		printf("        %-16s %s (only: ", all_opts[i].name,
 		       all_opts[i].desc);
 		for (j = 0; j < CS_ARCH_MAX; j++) {

--- a/cstool/cstool.c
+++ b/cstool/cstool.c
@@ -676,117 +676,117 @@ static void print_details(csh handle, cs_arch arch, cs_mode md, cs_insn *ins)
 
 	switch (arch) {
 	case CS_ARCH_X86:
-#if CAPSTONE_X86_SUPPORT
+#ifdef CAPSTONE_HAS_X86
 		print_insn_detail_x86(handle, md, ins);
 #endif
 		break;
 	case CS_ARCH_ARM:
-#if CAPSTONE_ARM_SUPPORT
+#ifdef CAPSTONE_HAS_ARM
 		print_insn_detail_arm(handle, ins);
 #endif
 		break;
 	case CS_ARCH_AARCH64:
-#if CAPSTONE_AARCH64_SUPPORT
+#ifdef CAPSTONE_HAS_AARCH64
 		print_insn_detail_aarch64(handle, ins);
 #endif
 		break;
 	case CS_ARCH_MIPS:
-#if CAPSTONE_MIPS_SUPPORT
+#ifdef CAPSTONE_HAS_MIPS
 		print_insn_detail_mips(handle, ins);
 #endif
 		break;
 	case CS_ARCH_PPC:
-#if CAPSTONE_PPC_SUPPORT
+#ifdef CAPSTONE_HAS_POWERPC
 		print_insn_detail_ppc(handle, ins);
 #endif
 		break;
 	case CS_ARCH_SPARC:
-#if CAPSTONE_SPARC_SUPPORT
+#ifdef CAPSTONE_HAS_SPARC
 		print_insn_detail_sparc(handle, ins);
 #endif
 		break;
 	case CS_ARCH_SYSTEMZ:
-#if CAPSTONE_SYSTEMZ_SUPPORT
+#ifdef CAPSTONE_HAS_SYSTEMZ
 		print_insn_detail_systemz(handle, ins);
 #endif
 		break;
 	case CS_ARCH_XCORE:
-#if CAPSTONE_XCORE_SUPPORT
+#ifdef CAPSTONE_HAS_XCORE
 		print_insn_detail_xcore(handle, ins);
 #endif
 		break;
 	case CS_ARCH_M68K:
-#if CAPSTONE_M68K_SUPPORT
+#ifdef CAPSTONE_HAS_M68K
 		print_insn_detail_m68k(handle, ins);
 #endif
 		break;
 	case CS_ARCH_TMS320C64X:
-#if CAPSTONE_TMS320C64X_SUPPORT
+#ifdef CAPSTONE_HAS_TMS320C64X
 		print_insn_detail_tms320c64x(handle, ins);
 #endif
 		break;
 	case CS_ARCH_M680X:
-#if CAPSTONE_M680X_SUPPORT
+#ifdef CAPSTONE_HAS_M680X
 		print_insn_detail_m680x(handle, ins);
 #endif
 		break;
 	case CS_ARCH_EVM:
-#if CAPSTONE_EVM_SUPPORT
+#ifdef CAPSTONE_HAS_EVM
 		print_insn_detail_evm(handle, ins);
 #endif
 		break;
 	case CS_ARCH_WASM:
-#if CAPSTONE_WASM_SUPPORT
+#ifdef CAPSTONE_HAS_WASM
 		print_insn_detail_wasm(handle, ins);
 #endif
 		break;
 	case CS_ARCH_MOS65XX:
-#if CAPSTONE_MOS65XX_SUPPORT
+#ifdef CAPSTONE_HAS_MOS65XX
 		print_insn_detail_mos65xx(handle, ins);
 #endif
 		break;
 	case CS_ARCH_BPF:
-#if CAPSTONE_BPF_SUPPORT
+#ifdef CAPSTONE_HAS_BPF
 		print_insn_detail_bpf(handle, ins);
 #endif
 		break;
 	case CS_ARCH_RISCV:
-#if CAPSTONE_RISCV_SUPPORT
+#ifdef CAPSTONE_HAS_RISCV
 		print_insn_detail_riscv(handle, ins);
 #endif
 		break;
 	case CS_ARCH_SH:
-#if CAPSTONE_SH_SUPPORT
+#ifdef CAPSTONE_HAS_SH
 		print_insn_detail_sh(handle, ins);
 #endif
 		break;
 	case CS_ARCH_TRICORE:
-#if CAPSTONE_TRICORE_SUPPORT
+#ifdef CAPSTONE_HAS_TRICORE
 		print_insn_detail_tricore(handle, ins);
 #endif
 		break;
 	case CS_ARCH_ALPHA:
-#if CAPSTONE_ALPHA_SUPPORT
+#ifdef CAPSTONE_HAS_ALPHA
 		print_insn_detail_alpha(handle, ins);
 #endif
 		break;
 	case CS_ARCH_HPPA:
-#if CAPSTONE_HPPA_SUPPORT
+#ifdef CAPSTONE_HAS_HPPA
 		print_insn_detail_hppa(handle, ins);
 #endif
 		break;
 	case CS_ARCH_LOONGARCH:
-#if CAPSTONE_LOONGARCH_SUPPORT
+#ifdef CAPSTONE_HAS_LOONGARCH
 		print_insn_detail_loongarch(handle, ins);
 #endif
 		break;
 	case CS_ARCH_XTENSA:
-#if CAPSTONE_XTENSA_SUPPORT
+#ifdef CAPSTONE_HAS_XTENSA
 		print_insn_detail_xtensa(handle, ins);
 #endif
 		break;
 	case CS_ARCH_ARC:
-#if CAPSTONE_ARC_SUPPORT
+#ifdef CAPSTONE_HAS_ARC
 		print_insn_detail_arc(handle, ins);
 #endif
 		break;

--- a/cstool/cstool.c
+++ b/cstool/cstool.c
@@ -676,73 +676,119 @@ static void print_details(csh handle, cs_arch arch, cs_mode md, cs_insn *ins)
 
 	switch (arch) {
 	case CS_ARCH_X86:
+#if CAPSTONE_X86_SUPPORT
 		print_insn_detail_x86(handle, md, ins);
+#endif
 		break;
 	case CS_ARCH_ARM:
+#if CAPSTONE_ARM_SUPPORT
 		print_insn_detail_arm(handle, ins);
+#endif
 		break;
 	case CS_ARCH_AARCH64:
+#if CAPSTONE_AARCH64_SUPPORT
 		print_insn_detail_aarch64(handle, ins);
+#endif
 		break;
 	case CS_ARCH_MIPS:
+#if CAPSTONE_MIPS_SUPPORT
 		print_insn_detail_mips(handle, ins);
+#endif
 		break;
 	case CS_ARCH_PPC:
+#if CAPSTONE_PPC_SUPPORT
 		print_insn_detail_ppc(handle, ins);
+#endif
 		break;
 	case CS_ARCH_SPARC:
+#if CAPSTONE_SPARC_SUPPORT
 		print_insn_detail_sparc(handle, ins);
+#endif
 		break;
 	case CS_ARCH_SYSTEMZ:
+#if CAPSTONE_SYSTEMZ_SUPPORT
 		print_insn_detail_systemz(handle, ins);
+#endif
 		break;
 	case CS_ARCH_XCORE:
+#if CAPSTONE_XCORE_SUPPORT
 		print_insn_detail_xcore(handle, ins);
+#endif
 		break;
 	case CS_ARCH_M68K:
+#if CAPSTONE_M68K_SUPPORT
 		print_insn_detail_m68k(handle, ins);
+#endif
 		break;
 	case CS_ARCH_TMS320C64X:
+#if CAPSTONE_TMS320C64X_SUPPORT
 		print_insn_detail_tms320c64x(handle, ins);
+#endif
 		break;
 	case CS_ARCH_M680X:
+#if CAPSTONE_M680X_SUPPORT
 		print_insn_detail_m680x(handle, ins);
+#endif
 		break;
 	case CS_ARCH_EVM:
+#if CAPSTONE_EVM_SUPPORT
 		print_insn_detail_evm(handle, ins);
+#endif
 		break;
 	case CS_ARCH_WASM:
+#if CAPSTONE_WASM_SUPPORT
 		print_insn_detail_wasm(handle, ins);
+#endif
 		break;
 	case CS_ARCH_MOS65XX:
+#if CAPSTONE_MOS65XX_SUPPORT
 		print_insn_detail_mos65xx(handle, ins);
+#endif
 		break;
 	case CS_ARCH_BPF:
+#if CAPSTONE_BPF_SUPPORT
 		print_insn_detail_bpf(handle, ins);
+#endif
 		break;
 	case CS_ARCH_RISCV:
+#if CAPSTONE_RISCV_SUPPORT
 		print_insn_detail_riscv(handle, ins);
+#endif
 		break;
 	case CS_ARCH_SH:
+#if CAPSTONE_SH_SUPPORT
 		print_insn_detail_sh(handle, ins);
+#endif
 		break;
 	case CS_ARCH_TRICORE:
+#if CAPSTONE_TRICORE_SUPPORT
 		print_insn_detail_tricore(handle, ins);
+#endif
 		break;
 	case CS_ARCH_ALPHA:
+#if CAPSTONE_ALPHA_SUPPORT
 		print_insn_detail_alpha(handle, ins);
+#endif
 		break;
 	case CS_ARCH_HPPA:
+#if CAPSTONE_HPPA_SUPPORT
 		print_insn_detail_hppa(handle, ins);
+#endif
 		break;
 	case CS_ARCH_LOONGARCH:
+#if CAPSTONE_LOONGARCH_SUPPORT
 		print_insn_detail_loongarch(handle, ins);
+#endif
 		break;
 	case CS_ARCH_XTENSA:
+#if CAPSTONE_XTENSA_SUPPORT
 		print_insn_detail_xtensa(handle, ins);
+#endif
 		break;
 	case CS_ARCH_ARC:
+#if CAPSTONE_ARC_SUPPORT
 		print_insn_detail_arc(handle, ins);
+#endif
 		break;
 	default:
 		break;

